app/lib/shared/utils.dart

```diff
@@ -361,3 +362,38 @@ bool fixedTimeEquals(String a, String b) {
   }
   return result == 0;
 }
+
+extension StreamBoundedForEach<T> on Stream<T> {
+  /// Executes items using [eachFn] with maximum [concurrency].
+  Future<void> boundedForEach(
+    int concurrency,
+    FutureOr<void> Function(T) eachFn,
+  ) async {
+    final pool = Pool(concurrency);
+    final futures = Queue<Future>();
+    // keeps the size of the futures limited, allows some difference in the run times
+    final maxQueueSize = concurrency * 2;
+    try {
+      await for (final item in this) {
+        final f = pool.withResource(() => eachFn(item));
+        futures.add(f);
+        while (futures.length > maxQueueSize) {
+          await futures.removeFirst();
+        }
+      }
+      await Future.wait(futures);
```

Comment:
If the first future throws, then this will never run and any subsequent future that throws will cause an uncaught exception, right?

Reply:
```
Future<void> main() async {
  try {
    await Future.wait(List.generate(
        10,
        (i) => Future.microtask(() {
              throw Exception('$i');
            })));
  } catch (e) {
    print('caught $e');
  }
}
```
Will catch the first exception, no further exception is uncaught. The `Future.wait` documentation says:

> If any future completes with an error, then the returned future completes with that error. If further futures also complete with errors, those errors are discarded.

Reply:
I think `discarded` means that errors are not unhandled.

Reply:
This is correct. Therefore, it is safe to use `Future.wait`, or am I reading something wrong?

Reply:
yes, but `await futures.removeFirst()` is not safe, and `Future.wait` might never be called.

Reply:
Updated to always use `Future.wait` with a simple condition. This is suboptimal for the concurrency, but maintaining exactly the given number of concurrently running functions requires more complex code, and we don't really need that level of optimization yet.